contracts/smart-account/modules/SessionValidationModules/ERC20SessionValidationModule.sol

```diff
@@ -79,4 +79,35 @@ contract ERC20SessionValidationModule is ISessionValidationModule {
                 _sessionKeySignature
             ) == sessionKey;
     }
+
+    /**
+     * @dev validates that the call (destinationContract, callValue, funcCallData)
+     * complies with the Session Key permissions represented by sessionKeyData
+     * @param destinationContract address of the contract to be called
+     * @param callValue value to be sent with the call
+     * @param _funcCallData the data for the call. is parsed inside the SVM
+     * @param _sessionKeyData SessionKey data, that describes sessionKey permissions
+     */
+    function validateSessionParams(
+        address destinationContract,
+        uint256 callValue,
+        bytes calldata _funcCallData,
+        bytes calldata _sessionKeyData
+    ) external virtual override {
+        (, address token, address recipient, uint256 maxAmount) = abi.decode(
+            _sessionKeyData,
+            (address, address, address, uint256)
+        );
+
+        require(destinationContract == token, "ERC20SV Invalid Token");
+        require(callValue == 0, "ERC20SV Non Zero Value");
+
+        (address recepientCalled, uint256 amount) = abi.decode(
+            _funcCallData[4:],
```

Comment:
what kind of actual decoding happens in SVM?

above same validation we had to check all of the below, 

```
{
            uint256 offset = uint256(bytes32(_op.callData[4 + 64:4 + 96]));
            uint256 length = uint256(
                bytes32(_op.callData[4 + offset:4 + offset + 32])
            );
            //we expect data to be the `IERC20.transfer(address, uint256)` calldata
            data = _op.callData[4 + offset + 32:4 + offset + 32 + length];
        }
        if (address(bytes20(data[16:36])) != recipient) {
            revert("ERC20SV Wrong Recipient");
        }
        if (uint256(bytes32(data[36:68])) > maxAmount) {
            revert("ERC20SV Max Amount Exceeded");
        }
```

Reply:
This code is extracting call parameters from the calldata and compares them to the permitted values.

>above same validation we had to check all of the below,

I'm not sure I fully get this

Reply:
you mentioned now this data comes from Router module so we don't parse it like earlier in SVM (code pasted by me above)
so I was trying to find similar parsing logic in router contract and could not find it.